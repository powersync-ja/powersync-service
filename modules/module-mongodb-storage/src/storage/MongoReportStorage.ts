import { mongo } from '@powersync/lib-service-mongodb';
import { storage } from '@powersync/service-core';
import { event_types } from '@powersync/service-types';
import { PowerSyncMongo } from './implementation/db.js';
import { SdkConnectDocument } from './implementation/models.js';

// import { SdkConnectDocument } from './implementation/models.js';
export class MongoReportStorage implements storage.ReportStorageFactory {
  private readonly client: mongo.MongoClient;
  public readonly db: PowerSyncMongo;

  constructor(db: PowerSyncMongo) {
    this.client = db.client;
    this.db = db;
  }

  async scrapeSdkData(data: event_types.PaginatedInstanceRequest): Promise<void> {
    console.log('MongoReportStorage.scrapeSdkData', data);
  }

  async reportSdkConnect(data: SdkConnectDocument): Promise<void> {
    await this.db.sdk_report_events.insertOne(data);
  }
  async reportSdkDisconnect(data: SdkConnectDocument): Promise<void> {
    const { _id, ...rest } = data;
    await this.db.sdk_report_events.findOneAndUpdate({ _id }, { $set: rest }, { upsert: true });
  }
  async listCurrentConnections(data: event_types.PaginatedInstanceRequest): Promise<void> {
    console.log('MongoReportStorage.listCurrentConnections', data);
  }

  async [Symbol.asyncDispose]() {
    // No-op
  }
}
