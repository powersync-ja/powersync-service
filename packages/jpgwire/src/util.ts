import * as tls from 'tls';
import { DEFAULT_CERTS } from './certs.js';
import * as pgwire from './pgwire.js';
import { PgType } from './pgwire_types.js';
import * as datefns from 'date-fns';

// TODO this is duplicated, but maybe that is ok
export interface NormalizedConnectionConfig {
  id: string;
  tag: string;

  hostname: string;
  port: number;
  database: string;

  username: string;
  password: string;

  sslmode: 'verify-full' | 'verify-ca' | 'disable';
  cacert: string | undefined;

  client_certificate: string | undefined;
  client_private_key: string | undefined;
}

type Mutable<T> = {
  -readonly [P in keyof T]: T[P];
};

export interface PgWireConnectionOptions extends NormalizedConnectionConfig {
  resolved_ip?: string;
}

export function clientTlsOptions(options: PgWireConnectionOptions): tls.ConnectionOptions {
  if (options.client_certificate && options.client_private_key) {
    return {
      cert: options.client_certificate,
      key: options.client_private_key
    };
  } else {
    return {};
  }
}
export function tlsOptions(options: PgWireConnectionOptions): false | tls.ConnectionOptions {
  if (options.sslmode == 'disable') {
    return false;
  } else if (options.sslmode == 'verify-full') {
    // Validate against default CA, or supplied CA.
    // Also validate hostname.
    return {
      ca: options.cacert ?? DEFAULT_CERTS,
      // hostname for TLS validation.
      // This may be different from the host we're connecting to if we pre-resolved
      // the IP.
      host: options.hostname,
      servername: options.hostname,
      ...clientTlsOptions(options)
    };
  } else if (options.sslmode == 'verify-ca') {
    if (options.cacert == null) {
      throw new Error(`cacert required for verify-ca`);
    }
    return {
      ca: options.cacert!,
      host: options.hostname,
      servername: options.hostname,
      // Disable hostname check
      checkServerIdentity: () => undefined,
      ...clientTlsOptions(options)
    };
  } else {
    throw new Error(`Unsupported sslmode: ${options.sslmode}`);
  }
}

export async function connectPgWire(config: PgWireConnectionOptions, options?: { type?: 'standard' | 'replication' }) {
  let connectionOptions: Mutable<ExtendedPgwireOptions> = {
    application_name: 'PowerSync',

    // tlsOptions below contains the original hostname
    hostname: config.resolved_ip ?? config.hostname,
    port: config.port,
    database: config.database,

    user: config.username,
    password: config.password
  };

  if (options?.type != 'standard') {
    connectionOptions.replication = 'database';
  }

  if (config.sslmode != 'disable') {
    connectionOptions.sslmode = 'require';

    // HACK: Not standard pgwire options
    // Just the easiest way to pass on our config to pgwire_node.js
    connectionOptions.sslrootcert = tlsOptions(config);
  } else {
    connectionOptions.sslmode = 'disable';
  }
  const connection = await pgwire.pgconnect(connectionOptions as pgwire.PgConnectOptions);
  // HACK: Replace row decoding with our own implementation
  (connection as any)._recvDataRow = _recvDataRow;
  return connection;
}

function _recvDataRow(this: any, _message: any, row: Uint8Array[], batch: any) {
  for (let i = 0; i < this._rowColumns.length; i++) {
    const valbuf = row[i];
    if (valbuf == null) {
      continue;
    }
    const { binary, typeOid } = this._rowColumns[i];
    // TODO avoid this._clientTextDecoder.decode for bytea
    row[i] = binary
      ? // do not valbuf.slice() because nodejs Buffer .slice does not copy
        // TODO but we not going to receive Buffer here ?
        Uint8Array.prototype.slice.call(valbuf)
      : PgType.decode(this._rowTextDecoder.decode(valbuf), typeOid);
  }
  batch.rows.push(row);
}

export interface PgPoolOptions {
  /**
   * Maximum number of open connections.
   *
   * Once this limit is reached, queries are queued.
   *
   * Defaults to 5.
   */
  maxSize?: number;

  /**
   * Idle timeout in ms before a connection is closed.
   */
  idleTimeout?: number | undefined;
}

/**
 * Open a postgres pool.
 *
 * The pool cannot be used for transactions, and cannot be used for logical replication.
 */
export function connectPgWirePool(config: PgWireConnectionOptions, options?: PgPoolOptions) {
  const idleTimeout = options?.idleTimeout;
  const maxSize = options?.maxSize ?? 5;

  let connectionOptions: Mutable<ExtendedPgwireOptions> = {
    application_name: 'PowerSync',

    // tlsOptions below contains the original hostname
    hostname: config.resolved_ip ?? config.hostname,
    port: config.port,
    database: config.database,

    user: config.username,
    password: config.password,

    _poolSize: maxSize,
    _poolIdleTimeout: idleTimeout
  };

  if (config.sslmode != 'disable') {
    connectionOptions.sslmode = 'require';

    // HACK: Not standard pgwire options
    // Just the easiest way to pass on our config to pgwire_node.js
    connectionOptions.sslrootcert = tlsOptions(config);
  } else {
    connectionOptions.sslmode = 'disable';
  }

  const pool = pgwire.pgpool(connectionOptions as pgwire.PgConnectOptions);
  const originalGetConnection = (pool as any)._getConnection;
  (pool as any)._getConnection = function (this: any) {
    const con = originalGetConnection.call(this);
    // HACK: Replace row decoding with our own implementation
    (con as any)._recvDataRow = _recvDataRow;
    return con;
  };
  return pool;
}

/**
 * Hack: sslrootcert is passed through as-is to pgwire_node.
 *
 * We use that to pass in custom TLS options, without having to modify pgwire itself.
 */
export interface ExtendedPgwireOptions extends pgwire.PgConnectKnownOptions {
  sslrootcert?: false | tls.ConnectionOptions;
}

export function lsnMakeComparable(text: string) {
  const [h, l] = text.split('/');
  return h.padStart(8, '0') + '/' + l.padStart(8, '0');
}

/**
 * Convert a postgres timestamptz to a SQLite-compatible/normalized timestamp.
 *
 * We output in the format:
 * YYYY-MM-DD HH:MM:SSZ
 * YYYY-MM-DD HH:MM:SS.SSSZ
 *
 * Precision is preserved from the source string, with the exception of ".000".
 *
 * We have specific exceptions for -infinity and infinity.
 */
export function timestamptzToSqlite(source?: string) {
  if (source == null) {
    return null;
  }
  // Make compatible with SQLite
  const match = /^([\d\-]+) ([\d:]+)(\.\d+)?([+-][\d:]+)$/.exec(source);
  if (match == null) {
    if (source == 'infinity') {
      return '9999-12-31 23:59:59Z';
    } else if (source == '-infinity') {
      return '0000-01-01 00:00:00Z';
    } else {
      return null;
    }
  }
  const [_, date, time, precision, timezone] = match as any;
  // datefns.parseISO() supports some dates that new Date() doesn't, e.g. '0022-09-08 21:19:28+00'
  const parsed = datefns.parseISO(`${date}T${time}${timezone}`);
  // Some dates are outside the range of JS and/or SQLite, e.g. 10022-09-08 21:19:28+00.
  // Just fallback to null for now.
  // We may consider preserving the source value in these cases.
  if (isNaN(parsed.getTime())) {
    return null;
  }
  const text = parsed.toISOString().replace('T', ' ').replace('.000', '').replace('Z', '');

  return `${text}${precision ?? ''}Z`;
}

/**
 * For timestamp without timezone, we keep it mostly as-is.
 *
 * We make specific exceptions for 'infinity' and '-infinity'.
 *
 * https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-DATETIME-SPECIAL-VALUES
 */
export function timestampToSqlite(source?: string) {
  if (source == null) {
    return null;
  }
  if (source == 'infinity') {
    return '9999-12-31 23:59:59';
  } else if (source == '-infinity') {
    return '0000-01-01 00:00:00';
  } else {
    return source;
  }
}
/**
 * For date, we keep it mostly as-is.
 *
 * We make specific exceptions for 'infinity' and '-infinity'.
 *
 * https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-DATETIME-SPECIAL-VALUES
 */
export function dateToSqlite(source?: string) {
  if (source == null) {
    return null;
  }
  if (source == 'infinity') {
    return '9999-12-31';
  } else if (source == '-infinity') {
    return '0000-01-01';
  } else {
    return source;
  }
}

/**
 * PgResult rows are arrays of values.
 *
 * This converts it to objects.
 */
export function pgwireRows(rs: pgwire.PgResult): Record<string, any>[] {
  const columns = rs.columns;
  return rs.rows.map((row) => {
    let r: Record<string, any> = {};
    for (let i = 0; i < columns.length; i++) {
      const c = columns[i];
      r[c.name] = row[i];
    }
    return r;
  });
}
