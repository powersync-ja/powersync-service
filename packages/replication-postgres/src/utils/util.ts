import * as pgwire from '@powersync/service-jpgwire';
import { replication } from '@powersync/service-core';

import { ReplicaIdentityResult } from '../types.js';
import { retriedQuery } from './pgwire_utils.js';

export function getReplicaIdColumns(relation: pgwire.PgoutputRelation): replication.ReplicationColumn[] {
  if (relation.replicaIdentity == 'nothing') {
    return [];
  } else {
    return relation.columns.filter((c) => (c.flags & 0b1) != 0).map((c) => ({ name: c.name, typeOid: c.typeOid }));
  }
}

export function getRelId(source: pgwire.PgoutputRelation): number {
  // Source types are wrong here
  const relId = (source as any).relationOid as number;
  if (relId == null || typeof relId != 'number') {
    throw new Error(`No relation id!`);
  }
  return relId;
}

export function getPgOutputRelation(source: pgwire.PgoutputRelation): replication.Relation {
  return {
    name: source.name,
    schema: source.schema,
    relationId: getRelId(source),
    replicaIdentity: source.replicaIdentity,
    replicationColumns: getReplicaIdColumns(source)
  };
}

export async function getPrimaryKeyColumns(
  db: pgwire.PgClient,
  relationId: number,
  mode: 'primary' | 'replident'
): Promise<replication.ReplicationColumn[]> {
  const indexFlag = mode == 'primary' ? `i.indisprimary` : `i.indisreplident`;
  const attrRows = await retriedQuery(db, {
    statement: `SELECT a.attname as name, a.atttypid as typeid, a.attnum as attnum
                                    FROM pg_index i
                                             JOIN pg_attribute a
                                                  ON a.attrelid = i.indrelid
                                                      AND a.attnum = ANY (i.indkey)
                                    WHERE i.indrelid = $1::oid
                                      AND ${indexFlag}
                                      AND a.attnum > 0
                                      ORDER BY a.attnum`,
    params: [{ value: relationId, type: 'int4' }]
  });

  return attrRows.rows.map((row) => {
    return { name: row[0] as string, typeOid: row[1] as number };
  });
}

export async function getAllColumns(db: pgwire.PgClient, relationId: number): Promise<replication.ReplicationColumn[]> {
  const attrRows = await retriedQuery(db, {
    statement: `SELECT a.attname as name, a.atttypid as typeid, a.attnum as attnum
                                    FROM pg_attribute a
                                    WHERE a.attrelid = $1::oid
                                      AND attnum > 0
                                    ORDER BY a.attnum`,
    params: [{ type: 'varchar', value: relationId }]
  });
  return attrRows.rows.map((row) => {
    return { name: row[0] as string, typeOid: row[1] as number };
  });
}

export async function getReplicationIdentityColumns(
  db: pgwire.PgClient,
  relationId: number
): Promise<ReplicaIdentityResult> {
  const rows = await retriedQuery(db, {
    statement: `SELECT CASE relreplident
        WHEN 'd' THEN 'default'
        WHEN 'n' THEN 'nothing'
        WHEN 'f' THEN 'full'
        WHEN 'i' THEN 'index'
     END AS replica_identity
FROM pg_class
WHERE oid = $1::oid LIMIT 1`,
    params: [{ type: 'int8', value: relationId }]
  });
  const idType: string = rows.rows[0]?.[0];
  if (idType == 'nothing' || idType == null) {
    return { replicationIdentity: 'nothing', columns: [] };
  } else if (idType == 'full') {
    return { replicationIdentity: 'full', columns: await getAllColumns(db, relationId) };
  } else if (idType == 'default') {
    return { replicationIdentity: 'default', columns: await getPrimaryKeyColumns(db, relationId, 'primary') };
  } else if (idType == 'index') {
    return { replicationIdentity: 'index', columns: await getPrimaryKeyColumns(db, relationId, 'replident') };
  } else {
    return { replicationIdentity: 'nothing', columns: [] };
  }
}
