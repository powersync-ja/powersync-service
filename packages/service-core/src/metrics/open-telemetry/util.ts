import { OTLPMetricExporter } from '@opentelemetry/exporter-metrics-otlp-http';
import { PrometheusExporter } from '@opentelemetry/exporter-prometheus';
import { MeterProvider, MetricReader, PeriodicExportingMetricReader } from '@opentelemetry/sdk-metrics';
import { logger } from '@powersync/lib-services-framework';
import { ServiceContext } from '../../system/ServiceContext.js';
import { MetricsFactory } from '../metrics-interfaces.js';
import { OpenTelemetryMetricsFactory } from './OpenTelemetryMetricsFactory.js';

import pkg from '../../../package.json' with { type: 'json' };
import { resourceFromAttributes } from '@opentelemetry/resources';

export function createOpenTelemetryMetricsFactory(context: ServiceContext): MetricsFactory {
  const { configuration, lifeCycleEngine, storageEngine } = context;
  const configuredExporters: MetricReader[] = [];

  if (configuration.telemetry.prometheus_port) {
    const prometheusExporter = new PrometheusExporter({
      port: configuration.telemetry.prometheus_port,
      preventServerStart: true
    });
    configuredExporters.push(prometheusExporter);

    lifeCycleEngine.withLifecycle(prometheusExporter, {
      start: async () => {
        await prometheusExporter.startServer();
        logger.info(`Prometheus metric export enabled on port:${configuration.telemetry.prometheus_port}`);
      }
    });
  }

  if (!configuration.telemetry.disable_telemetry_sharing) {
    const periodicExporter = new PeriodicExportingMetricReader({
      exporter: new OTLPMetricExporter({
        url: configuration.telemetry.internal_service_endpoint
      }),
      exportIntervalMillis: 1000 * 60 * 5 // 5 minutes
    });

    configuredExporters.push(periodicExporter);
  }

  let resolvedInstanceId: (id: string) => void;
  const instanceIdPromise = new Promise<string>((resolve) => {
    resolvedInstanceId = resolve;
  });

  lifeCycleEngine.withLifecycle(null, {
    start: async () => {
      const bucketStorage = storageEngine.activeBucketStorage;
      try {
        const instanceId = await bucketStorage.getPowerSyncInstanceId();
        resolvedInstanceId(instanceId);
      } catch (err) {
        resolvedInstanceId('Unknown');
      }
    }
  });

  const resource = resourceFromAttributes({
    ['service']: 'PowerSync',
    ['service.version']: pkg.version,
    ['instance_id']: instanceIdPromise
  });

  // This triggers OpenTelemetry to resolve the async attributes (instanceIdPromise).
  // This will never reject, and we don't specifically need to wait for it.
  resource.waitForAsyncAttributes?.();

  const meterProvider = new MeterProvider({
    resource,

    readers: configuredExporters
  });

  lifeCycleEngine.withLifecycle(meterProvider, {
    stop: async () => {
      await meterProvider.shutdown();
    }
  });

  const meter = meterProvider.getMeter('powersync');

  return new OpenTelemetryMetricsFactory(meter);
}
