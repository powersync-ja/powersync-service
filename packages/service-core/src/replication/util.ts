import * as pgwire from '@powersync/service-jpgwire';

import * as util from '../util/util-index.js';
import { ReplicationColumn, ReplicationIdentity } from './PgRelation.js';

export interface ReplicaIdentityResult {
  columns: ReplicationColumn[];
  replicationIdentity: ReplicationIdentity;
}

export async function getPrimaryKeyColumns(
  db: pgwire.PgClient,
  relationId: number,
  mode: 'primary' | 'replident'
): Promise<ReplicationColumn[]> {
  const indexFlag = mode == 'primary' ? `i.indisprimary` : `i.indisreplident`;
  const attrRows = await util.retriedQuery(db, {
    statement: `SELECT a.attname as name, a.atttypid as typeid, a.attnum as attnum
                                    FROM pg_index i
                                             JOIN pg_attribute a
                                                  ON a.attrelid = i.indrelid
                                                      AND a.attnum = ANY (i.indkey)
                                    WHERE i.indrelid = $1::oid
                                      AND ${indexFlag}
                                      AND a.attnum > 0
                                      ORDER BY a.attnum`,
    params: [{ value: relationId, type: 'int4' }]
  });

  return attrRows.rows.map((row) => {
    return { name: row[0] as string, typeOid: row[1] as number };
  });
}

export async function getAllColumns(db: pgwire.PgClient, relationId: number): Promise<ReplicationColumn[]> {
  const attrRows = await util.retriedQuery(db, {
    statement: `SELECT a.attname as name, a.atttypid as typeid, a.attnum as attnum
                                    FROM pg_attribute a
                                    WHERE a.attrelid = $1::oid
                                      AND attnum > 0
                                    ORDER BY a.attnum`,
    params: [{ type: 'varchar', value: relationId }]
  });
  return attrRows.rows.map((row) => {
    return { name: row[0] as string, typeOid: row[1] as number };
  });
}

export async function getReplicationIdentityColumns(
  db: pgwire.PgClient,
  relationId: number
): Promise<ReplicaIdentityResult> {
  const rows = await util.retriedQuery(db, {
    statement: `SELECT CASE relreplident
        WHEN 'd' THEN 'default'
        WHEN 'n' THEN 'nothing'
        WHEN 'f' THEN 'full'
        WHEN 'i' THEN 'index'
     END AS replica_identity
FROM pg_class
WHERE oid = $1::oid LIMIT 1`,
    params: [{ type: 'int8', value: relationId }]
  });
  const idType: string = rows.rows[0]?.[0];
  if (idType == 'nothing' || idType == null) {
    return { replicationIdentity: 'nothing', columns: [] };
  } else if (idType == 'full') {
    return { replicationIdentity: 'full', columns: await getAllColumns(db, relationId) };
  } else if (idType == 'default') {
    return { replicationIdentity: 'default', columns: await getPrimaryKeyColumns(db, relationId, 'primary') };
  } else if (idType == 'index') {
    return { replicationIdentity: 'index', columns: await getPrimaryKeyColumns(db, relationId, 'replident') };
  } else {
    return { replicationIdentity: 'nothing', columns: [] };
  }
}
