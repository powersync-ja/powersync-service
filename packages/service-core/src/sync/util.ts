import * as timers from 'timers/promises';

import * as util from '../util/util-index.js';
import { Metrics } from '../metrics/Metrics.js';

export type TokenStreamOptions = {
  /**
   * Adds periodic keepalive events
   */
  keep_alive: boolean;
  /**
   * Warn before the token is going to expire
   */
  expire_warning_period: number;
};

const KEEPALIVE_INTERVAL = 20_000;

const DEFAULT_TOKEN_STREAM_OPTIONS: TokenStreamOptions = {
  keep_alive: true,
  expire_warning_period: 20_000
};

/**
 * An iterator that periodically yields token and optionally keepalive events, and returns once the
 * provided token expiry is reached.
 *
 * @param token exp is expiry as a unix timestamp (seconds)
 * @param signal abort the iterator with this
 * @param options configure keepalive and expire warnings
 */
export async function* tokenStream(
  token: { exp: number },
  signal: AbortSignal,
  options?: Partial<TokenStreamOptions>
): AsyncGenerator<util.StreamingSyncKeepalive> {
  const resolved_options: TokenStreamOptions = {
    ...DEFAULT_TOKEN_STREAM_OPTIONS,
    ...(options ?? {})
  };

  const { keep_alive, expire_warning_period } = resolved_options;

  // Real tokens always have an integer for exp.
  // In tests, we may use fractional seconds to reduce the delay.
  // Both cases are handled here.
  const expires_at = token.exp * 1000;
  const expire_warning_at = expires_at - expire_warning_period;

  let first_expire_test = true;

  while (!signal.aborted) {
    const token_expires_in = Math.max(0, Math.ceil(token.exp - Date.now() / 1000));
    if (first_expire_test && token_expires_in > 0) {
      first_expire_test = false;
    } else {
      yield { token_expires_in: token_expires_in };
      if (token_expires_in == 0) {
        return;
      }
    }

    const keep_alive_delay = KEEPALIVE_INTERVAL * (0.9 + Math.random() * 0.2);

    // Add margin due to setTimeout inaccuracies
    const expiry_delay = Math.max(0, expires_at - Date.now() + 3);
    const expiry_warning_delay = Math.max(0, expire_warning_at - Date.now() + 3);
    // Either the warning has past or it's before
    const relevant_expiry_delay = expiry_warning_delay != 0 ? expiry_warning_delay : expiry_delay;

    const delay = keep_alive ? Math.min(relevant_expiry_delay, keep_alive_delay) : relevant_expiry_delay;
    await timers.setTimeout(delay, null, { signal }).catch(() => {
      // Ignore AbortError
    });
  }
}

export async function* ndjson(iterator: AsyncIterable<string | null | Record<string, any>>): AsyncGenerator<string> {
  for await (let data of iterator) {
    if (data == null) {
      // Empty value to flush iterator memory
      continue;
    } else if (typeof data == 'string') {
      // Pre-serialized value
      yield data + '\n';
    } else {
      yield JSON.stringify(data) + '\n';
    }
  }
}

export async function* transformToBytesTracked(iterator: AsyncIterable<string>): AsyncGenerator<Buffer> {
  for await (let data of iterator) {
    const encoded = Buffer.from(data, 'utf8');
    Metrics.getInstance().data_synced_bytes.add(encoded.length);
    yield encoded;
  }
}
