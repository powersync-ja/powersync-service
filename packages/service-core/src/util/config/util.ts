import * as urijs from 'uri-js';

export function getSupabaseJwksUrl(connection: any) {
  if (connection == null) {
    return null;
  } else if (connection.type != 'postgresql') {
    return null;
  }

  let hostname: string | undefined = connection.hostname;
  if (hostname == null && typeof connection.uri == 'string') {
    hostname = urijs.parse(connection.uri).host;
  }
  if (hostname == null) {
    return null;
  }

  const match = /db.(\w+).supabase.co/.exec(hostname);
  if (match == null) {
    return null;
  }
  const projectId = match[1];

  return `https://${projectId}.supabase.co/auth/v1/.well-known/jwks.json`;
}
