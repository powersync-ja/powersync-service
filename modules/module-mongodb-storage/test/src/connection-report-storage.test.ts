import { afterAll, beforeAll, describe, expect, it } from 'vitest';
import { INITIALIZED_MONGO_REPORT_STORAGE_FACTORY } from './util.js';
import { register, ReportUserData } from '@powersync/service-core-tests';
import { event_types } from '@powersync/service-types';
import { MongoReportStorage } from '@module/storage/MongoReportStorage.js';

const userData = register.REPORT_TEST_USERS;
const dates = register.REPORT_TEST_DATES;
const factory = await INITIALIZED_MONGO_REPORT_STORAGE_FACTORY();

function removeVolatileFields(
  connections: event_types.ClientConnection[]
): Partial<event_types.ClientConnection & { _id: string }>[] {
  return connections.map((sdk: Partial<event_types.ClientConnection & { _id: string }>) => {
    const { _id, disconnected_at, connected_at, jwt_exp, ...rest } = sdk;
    return {
      ...rest
    };
  });
}

async function loadData(data: ReportUserData, factory: MongoReportStorage) {
  await factory.db.connection_report_events.insertMany(Object.values(data));
}

async function deleteData(factory: MongoReportStorage) {
  await factory.db.connection_report_events.deleteMany();
}

beforeAll(async () => {
  await loadData(userData, factory);
});
afterAll(async () => {
  await deleteData(factory);
});

describe('Report storage tests', async () => {
  await register.registerReportTests(factory);
});

describe('Connection reporting storage', async () => {
  it('Should create a connection report if its after a day', async () => {
    const newConnectAt = new Date(
      dates.now.getFullYear(),
      dates.now.getMonth(),
      dates.now.getDate() + 1,
      dates.now.getHours()
    );
    const jwtExp = new Date(newConnectAt.getFullYear(), newConnectAt.getMonth(), newConnectAt.getDate() + 1);

    await factory.reportClientConnection({
      sdk: userData.user_week.sdk,
      connected_at: newConnectAt,
      jwt_exp: jwtExp,
      client_id: userData.user_week.client_id,
      user_id: userData.user_week.user_id,
      user_agent: userData.user_week.user_agent
    });

    const connection = await factory.db.connection_report_events.find({ user_id: userData.user_week.user_id }).toArray();
    expect(connection).toHaveLength(2);
    const cleaned = removeVolatileFields(connection);
    expect(cleaned).toMatchSnapshot();
  });

  it('Should update a connection report if its within a day', async () => {
    const newConnectAt = new Date(
      dates.now.getFullYear(),
      dates.now.getMonth(),
      dates.now.getDate(),
      dates.now.getHours(),
      dates.now.getMinutes() + 20
    );
    const jwtExp = new Date(newConnectAt.getFullYear(), newConnectAt.getMonth(), newConnectAt.getDate() + 1);
    await factory.reportClientConnection({
      sdk: userData.user_one.sdk,
      connected_at: newConnectAt,
      jwt_exp: jwtExp,
      client_id: userData.user_one.client_id,
      user_id: userData.user_one.user_id,
      user_agent: userData.user_one.user_agent
    });

    const connection = await factory.db.connection_report_events
      .find({ user_id: userData.user_one.user_id, client_id: userData.user_one.client_id })
      .toArray();
    expect(connection).toHaveLength(1);
    expect(new Date(connection[0].connected_at)).toEqual(newConnectAt);
    expect(new Date(connection[0].jwt_exp!)).toEqual(jwtExp);
    expect(connection[0].disconnected_at).toBeUndefined();
    const cleaned = removeVolatileFields(connection);
    expect(cleaned).toMatchSnapshot();
  });

  it('Should update a connected connection report and make it disconnected', async () => {
    const disconnectAt = new Date(
      dates.now.getFullYear(),
      dates.now.getMonth(),
      dates.now.getDate(),
      dates.now.getHours(),
      dates.now.getMinutes() + 20
    );
    const jwtExp = new Date(disconnectAt.getFullYear(), disconnectAt.getMonth(), disconnectAt.getDate() + 1);

    await factory.reportClientDisconnection({
      disconnected_at: disconnectAt,
      jwt_exp: jwtExp,
      client_id: userData.user_three.client_id,
      user_id: userData.user_three.user_id,
      user_agent: userData.user_three.user_agent,
      connected_at: userData.user_three.connected_at
    });

    const connection = await factory.db.connection_report_events.find({ user_id: userData.user_three.user_id }).toArray();
    expect(connection).toHaveLength(1);
    expect(new Date(connection[0].disconnected_at!)).toEqual(disconnectAt);
    const cleaned = removeVolatileFields(connection);
    expect(cleaned).toMatchSnapshot();
  });

  it('Should delete rows older than specified range', async () => {
    await deleteData(factory);
    await loadData(userData, factory);
    await factory.deleteOldConnectionData({
      date: dates.weekAgo
    });
    const connection = await factory.getClientConnectionReports({
      start: dates.monthAgo,
      end: dates.now
    });
    expect(connection).toMatchSnapshot();
  });
});
