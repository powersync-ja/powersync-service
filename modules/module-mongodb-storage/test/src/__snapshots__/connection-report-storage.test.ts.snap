// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SDK reporting storage > Should create a connection report if its after a day 1`] = `
[
  {
    "client_id": "client_week",
    "sdk": "powersync-js/1.24.5",
    "user_agent": "powersync-js/1.21.0 powersync-web Firefox/141 linux",
    "user_id": "user_week",
  },
  {
    "client_id": "client_week",
    "sdk": "powersync-js/1.24.5",
    "user_agent": "powersync-js/1.21.0 powersync-web Firefox/141 linux",
    "user_id": "user_week",
  },
]
`;

exports[`SDK reporting storage > Should delete rows older than specified range 1`] = `
{
  "sdks": [
    {
      "clients": 1,
      "sdk": "powersync-dart/1.6.4",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.21.1",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.21.2",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.21.4",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.24.5",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "unknown",
      "users": 1,
    },
  ],
  "users": 5,
}
`;

exports[`SDK reporting storage > Should show connected users with start range 1`] = `
{
  "sdks": [
    {
      "clients": 1,
      "sdk": "powersync-dart/1.6.4",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.21.1",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "unknown",
      "users": 1,
    },
  ],
  "users": 2,
}
`;

exports[`SDK reporting storage > Should show connected users with start range and end range 1`] = `
{
  "sdks": [
    {
      "clients": 1,
      "sdk": "powersync-js/1.21.1",
      "users": 1,
    },
  ],
  "users": 1,
}
`;

exports[`SDK reporting storage > Should show connection report data for user over the past day 1`] = `
{
  "sdks": [
    {
      "clients": 1,
      "sdk": "powersync-dart/1.6.4",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.21.1",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.21.4",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "unknown",
      "users": 1,
    },
  ],
  "users": 3,
}
`;

exports[`SDK reporting storage > Should show connection report data for user over the past month 1`] = `
{
  "sdks": [
    {
      "clients": 1,
      "sdk": "powersync-dart/1.6.4",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.21.1",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.21.2",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.21.4",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.23.6",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.23.7",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.24.5",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "unknown",
      "users": 1,
    },
  ],
  "users": 7,
}
`;

exports[`SDK reporting storage > Should show connection report data for user over the past week 1`] = `
{
  "sdks": [
    {
      "clients": 1,
      "sdk": "powersync-dart/1.6.4",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.21.1",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.21.2",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.21.4",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.24.5",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "unknown",
      "users": 1,
    },
  ],
  "users": 5,
}
`;

exports[`SDK reporting storage > Should update a connected connection report and make it disconnected 1`] = `
[
  {
    "client_id": "client_three",
    "sdk": "powersync-js/1.21.2",
    "user_agent": "powersync-js/1.21.0 powersync-web Firefox/141 linux",
    "user_id": "user_three",
  },
]
`;

exports[`SDK reporting storage > Should update a connection report if its within a day 1`] = `
[
  {
    "client_id": "client_one",
    "sdk": "powersync-dart/1.6.4",
    "user_agent": "powersync-dart/1.6.4 Dart (flutter-web) Chrome/128 android",
    "user_id": "user_one",
  },
]
`;
